IndexedDB: Add tests for HTML monkey patches

inexorabletash opened the pull request:
The IDB spec implicitly "monkey patches" HTML with imprecise prose
when describing newly created transactions:

"When control is returned to the event loop, the implementation must
unset the active flag."

The plan is replace that with a proper hook. Adding tests first to ensure the
expected behavior is captured appropriately.

https://github.com/w3c/IndexedDB/issues/87

inexorabletash commented:
All tests pass in Chrome 55 (stable)
The "fire-event..." tests all pass in FF 50. The other two have failures, but should pass "once microtask handling around Promises is fixed" - @smaug---- - could you take a peek?

w3c-bots commented:
# Chrome (unstable channel) #
Testing web-platform-tests at revision 2613e0d245767eceefb9ac1509f8916ffd9a98a8
Using browser at version 58.0.3013.3 dev
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary>5 tests ran</summary>
<details>
<summary><a href="http://w3c-test.org/submissions/4643/IndexedDB/fire-error-event-exception.html">/IndexedDB/fire-error-event-exception.html</a></summary>


|                                Subtest                                 | Results |
|------------------------------------------------------------------------|---------|
|                                                                        | OK      |
| `Exception in error event listener on request`                         | PASS    |
| `Exception in error event listener on transaction`                     | PASS    |
| `Exception in error event handler on request, with preventDefault`     | PASS    |
| `Exception in second error event listener on connection`               | PASS    |
| `Exception in error event handler on connection, with preventDefault`  | PASS    |
| `Exception in error event listener on connection`                      | PASS    |
| `Exception in error event handler on connection`                       | PASS    |
| `Exception in second error event listener on transaction`              | PASS    |
| `Exception in second error event listener on request`                  | PASS    |
| `Exception in error event handler on transaction`                      | PASS    |
| `Exception in error event handler on transaction, with preventDefault` | PASS    |
| `Exception in error event handler on request`                          | PASS    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4643/IndexedDB/fire-success-event-exception.html">/IndexedDB/fire-success-event-exception.html</a></summary>


|                         Subtest                         | Results |
|---------------------------------------------------------|---------|
|                                                         | OK      |
| `Exception in second success event listener on request` | PASS    |
| `Exception in success event handler on request`         | PASS    |
| `Exception in success event listener on request`        | PASS    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4643/IndexedDB/transaction-deactivation-timing.html">/IndexedDB/transaction-deactivation-timing.html</a></summary>


|                                       Subtest                                       | Results |
|-------------------------------------------------------------------------------------|---------|
|                                                                                     | OK      |
| `Deactivation of new transactions happens at end of invocation`                     | PASS    |
| `New transactions are not deactivated until after the microtask checkpoint`         | PASS    |
| `New transactions are deactivated before next task`                                 | PASS    |
| `New transactions from microtask are still active through the microtask checkpoint` | PASS    |
| `New transactions from microtask are deactivated before next task`                  | PASS    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4643/IndexedDB/event-dispatch-active-flag.html">/IndexedDB/event-dispatch-active-flag.html</a></summary>


|                      Subtest                       | Results |
|----------------------------------------------------|---------|
|                                                    | OK      |
| `Transactions are active during error handlers`    | PASS    |
| `Transactions are active during success listeners` | PASS    |
| `Transactions are active during error listeners`   | PASS    |
| `Transactions are active during success handlers`  | PASS    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4643/IndexedDB/fire-upgradeneeded-event-exception.html">/IndexedDB/fire-upgradeneeded-event-exception.html</a></summary>


|                   Subtest                    | Results |
|----------------------------------------------|---------|
|                                              | OK      |
| `Exception in upgradeneeded listener`        | PASS    |
| `Exception in upgradeneeded handler`         | PASS    |
| `Exception in second upgradeneeded listener` | PASS    |

</details>
</details>

w3c-bots commented:
# Firefox (nightly channel) #
Testing web-platform-tests at revision 2613e0d245767eceefb9ac1509f8916ffd9a98a8
Using browser at version BuildID 20170123125947; SourceStamp 36486fdc3813ef7943ae5b07b4128866d1938a6c
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary>5 tests ran</summary>
<details>
<summary><a href="http://w3c-test.org/submissions/4643/IndexedDB/fire-error-event-exception.html">/IndexedDB/fire-error-event-exception.html</a></summary>


|                                Subtest                                 | Results |
|------------------------------------------------------------------------|---------|
|                                                                        | OK      |
| `Exception in error event listener on request`                         | PASS    |
| `Exception in error event listener on transaction`                     | PASS    |
| `Exception in error event handler on request, with preventDefault`     | PASS    |
| `Exception in second error event listener on connection`               | PASS    |
| `Exception in error event handler on connection, with preventDefault`  | PASS    |
| `Exception in error event listener on connection`                      | PASS    |
| `Exception in error event handler on connection`                       | PASS    |
| `Exception in second error event listener on transaction`              | PASS    |
| `Exception in second error event listener on request`                  | PASS    |
| `Exception in error event handler on transaction`                      | PASS    |
| `Exception in error event handler on transaction, with preventDefault` | PASS    |
| `Exception in error event handler on request`                          | PASS    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4643/IndexedDB/fire-success-event-exception.html">/IndexedDB/fire-success-event-exception.html</a></summary>


|                         Subtest                         | Results |
|---------------------------------------------------------|---------|
|                                                         | OK      |
| `Exception in second success event listener on request` | PASS    |
| `Exception in success event handler on request`         | PASS    |
| `Exception in success event listener on request`        | PASS    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4643/IndexedDB/transaction-deactivation-timing.html">/IndexedDB/transaction-deactivation-timing.html</a></summary>


|                                       Subtest                                       | Results |
|-------------------------------------------------------------------------------------|---------|
|                                                                                     | OK      |
| `Deactivation of new transactions happens at end of invocation`                     | FAIL    |
| `New transactions are not deactivated until after the microtask checkpoint`         | FAIL    |
| `New transactions are deactivated before next task`                                 | PASS    |
| `New transactions from microtask are still active through the microtask checkpoint` | FAIL    |
| `New transactions from microtask are deactivated before next task`                  | PASS    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4643/IndexedDB/event-dispatch-active-flag.html">/IndexedDB/event-dispatch-active-flag.html</a></summary>


|                      Subtest                       | Results |
|----------------------------------------------------|---------|
|                                                    | OK      |
| `Transactions are active during error handlers`    | FAIL    |
| `Transactions are active during success listeners` | FAIL    |
| `Transactions are active during error listeners`   | FAIL    |
| `Transactions are active during success handlers`  | FAIL    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4643/IndexedDB/fire-upgradeneeded-event-exception.html">/IndexedDB/fire-upgradeneeded-event-exception.html</a></summary>


|                   Subtest                    | Results |
|----------------------------------------------|---------|
|                                              | OK      |
| `Exception in upgradeneeded listener`        | PASS    |
| `Exception in upgradeneeded handler`         | PASS    |
| `Exception in second upgradeneeded listener` | PASS    |

</details>
</details>

w3c-bots commented:
These tests are now available on [w3c-test.org](https://w3c-test.org/submissions/4643)

inexorabletash commented:
I split this up - this now covers just the transaction activation cases, i.e. those discussed in https://github.com/w3c/IndexedDB/issues/87 that would require the HTML hook as discussed in https://github.com/whatwg/html/pull/2310

domenic commented:
Do you have a reviewer lined up for these, or should @annevk and I try to muddle through based on our IDB knowledge? :)

(Also, needs a bit of a rebase, I guess.)

inexorabletash commented:
I'll wrangle an IDB-knowledgeable reviewer.

inexorabletash committed:
IndexedDB: Add tests transaction deactivation

The IDB spec implicitly "monkey patches" HTML with imprecise prose
when describing newly created transactions:

"When control is returned to the event loop, the implementation must
unset the active flag."

The plan is replace that with a proper hook. Adding tests first to
ensure that the expected behavior is captured appropriately.

https://github.com/w3c/IndexedDB/issues/87

w3c-bots commented:
These tests are now available on [w3c-test.org](https://w3c-test.org/submissions/4643)

inexorabletash committed:
Factor out helper

inexorabletash commented:
@pwnall - can you review?

pwnall commented on IndexedDB/event-dispatch-active-flag.html:
```diff
@@ -0,0 +1,153 @@
+  },
+  (t, db) => {
+    const tx = db.transaction('store');
+    const finish = keep_alive(tx, 'store');
```
finish sounds a bit close to done. How about a name like `release_tx`?

inexorabletash replied:
Done

pwnall commented on IndexedDB/event-dispatch-active-flag.html:
```diff
@@ -0,0 +1,153 @@
+    db.createObjectStore('store');
+  },
+  (t, db) => {
+    const tx = db.transaction('store');
```
Can you add a `oncomplete` handler and assert that it hasn't been called on line 38 (where you call finish)? Alternatively, this can all be done in `keep_alive`, but it's less explicit. An assert on line 38 would be a nice and easy way for a reader to make sure that the transaction is still pinned, as expected.

inexorabletash replied:
Done

pwnall commented on IndexedDB/event-dispatch-active-flag.html:
```diff
@@ -0,0 +1,153 @@
+      Promise.resolve().then(t.step_func(() => {
+        saw_handler_promise = true;
+        assert_true(is_transaction_active(tx, 'store'),
+                    'Transaction should be active in handlers microtasks');
```
handler microtasks?

inexorabletash replied:
should be "handler's microtasks"

pwnall replied:
Agreed. I'd be ok with `handler microtasks` if you don't want to change the outer quotes. I think the grammar still parses correctly, even though the form you suggested definitely has the lightest cognitive load.

inexorabletash replied:
I'll add the apostrophe.

pwnall commented on IndexedDB/event-dispatch-active-flag.html:
```diff
@@ -0,0 +1,153 @@
+      Promise.resolve().then(t.step_func(() => {
+        saw_listener_promise = true;
+        assert_true(is_transaction_active(tx, 'store'),
+                    'Transaction should be active in listeners microtasks');
```
listener microtasks?

inexorabletash replied:
Ditto.

pwnall commented on IndexedDB/transaction-deactivation-timing.html:
```diff
@@ -0,0 +1,144 @@
+
+      Promise.resolve().then(t.step_func(() => {
+        microtasks_ran = true;
+        new_tx = db.transaction('store');
```
Did you mean to reassign `new_tx`?

inexorabletash replied:
Whooops! Nope.

inexorabletash committed:
Incorporate review feedback

inexorabletash commented:
Thanks @pwnall - another look pls?

inexorabletash committed:
Rename finish to release_tx in other test file

inexorabletash committed:
Add transaction activation tests for upgrade transactions

pwnall commented on IndexedDB/transaction-deactivation-timing.html:
```diff
@@ -0,0 +1,143 @@
+    setTimeout(t.step_func(() => {
+      assert_false(is_transaction_active(tx, 'store'),
+                   'Transaction should be inactive in next task');
+      release_tx();
```
This could also benefit from a `tx_completed` check. I think that's generally the case when you check that the transaction should be inactive.

inexorabletash replied:
Thanks. I ended up just moving it into the `keep_alive()` helper since it was getting copy/pastey

pwnall reviewed (approved):
LGTM with one nit.

inexorabletash committed:
Factor tx complete test into keep_alive helper
